language/tools/move-cli/tests/cli_testsuite.rs

```diff
@@ -0,0 +1,144 @@
+/// runs `move clean`.
+/// This is designed for testing the CLI itself, but it can also be used to write
+/// tests for user-defined scripts/modules.
+
+pub const CLI_BINARY: &str = "../../../target/debug/move-cli";
+pub const CLI_TESTSUITE_DIR: &str = "tests/testsuite";
+
+const EXP_EXT: &str = "exp";
+
+// if this env var is set, `move clean` will not be run after each test.
+// this is useful if you want to look at the `move_data` produced by
+// a test.
+const NO_MOVE_CLEAN: &str = "NO_MOVE_CLEAN";
+// if either of these env vars is set, the test harness overwrites the
+// old .exp files with the output instead of checking them against the
+// output.
+const UPDATE_BASELINE: &str = "UPDATE_BASELINE";
+const UB: &str = "UB";
+
+fn format_diff(expected: String, actual: String) -> String {
+    use difference::*;
+
+    let changeset = Changeset::new(&expected, &actual, "\n");
+
+    let mut ret = String::new();
+
+    for seq in changeset.diffs {
+        match &seq {
+            Difference::Same(x) => {
+                ret.push_str(x);
+                ret.push_str("\n");
+            }
+            Difference::Add(x) => {
+                ret.push_str("\x1B[92m");
+                ret.push_str(x);
+                ret.push_str("\x1B[0m");
+                ret.push_str("\n");
+            }
+            Difference::Rem(x) => {
+                ret.push_str("\x1B[91m");
+                ret.push_str(x);
+                ret.push_str("\x1B[0m");
+                ret.push_str("\n");
+            }
+        }
+    }
+    ret
+}
+
+// Runs all tests under the test/testsuite directory
+fn cli_testsuite(args_path: &Path) -> datatest_stable::Result<()> {
+    let args_file = io::BufReader::new(File::open(args_path)?).lines();
+    // path where we will run the binary
+    let exe_dir = args_path.parent().unwrap();
+    let cli_binary_path = Path::new(CLI_BINARY).canonicalize()?;
+    let move_data = Path::new(exe_dir).join(MOVE_DATA);
+    let move_src = Path::new(exe_dir).join(MOVE_SRC);
+    let move_src_exists_before = move_src.exists();
+    assert!(
+        !move_data.exists(),
+        "tests should never include a {:?} directory",
+        MOVE_DATA
+    );
+    let mut output = "".to_string();
+    for args_line in args_file {
+        let args_line = args_line?;
+        if args_line.starts_with('#') {
+            // allow comments in args.txt
+            continue;
+        }
+        let args_iter: Vec<&str> = args_line.split_whitespace().collect();
+        if args_iter.is_empty() {
+            // allow blank lines
+            continue;
+        }
+        let cmd_output = Command::new(cli_binary_path.clone())
+            .current_dir(exe_dir)
+            .args(args_iter)
+            .output()?;
+        output += &format!("Command `{}`:\n", args_line);
+        output += std::str::from_utf8(&cmd_output.stdout)?;
+        output += std::str::from_utf8(&cmd_output.stderr)?;
+    }
```

Comment:
This is rather cool. Might be worth breaking it out into a separate entry point in `move_cli`. Could be really neat to easily setup batched commands in the CLI and reuse them easily

Reply:
Ooh, I like this idea a lot-- a `move test` command would be very cool/useful for folks that want to write expected value tests for their custom modules/scripts. We could probably rewrite a lot of our Move VM functional tests using this (CC @vgao1996).

Another thing I'd like to to is make it possible to run the batched commands in the same process + redirect stderr/stdout to a local buffer instead of spawning a new process for every command in the batch (which is slow).

Reply:
> We could probably rewrite a lot of our Move VM functional tests using this

Great idea. Functional tests were a copy of LLVM's tests and this is exactly how things work there. Although there was a very specific reason I didn't choose invoke the binaries at the beginning which I can't remember now... Anyways I'll review this and see how we'll make things work. In general I really like the idea of using the cli for testing or make the new move testing flow part of the cli.

Reply:
I like this so much I already added preliminary support for it (see `test` subcommand + test.rs file, which contains most of what used to be in cli_testsuite.rs). Some work needs to be done to make `move test` support running a whole directory of tests instead of a single args.txt file (I couldn't figure out the datatest magic to make this work), but should be possible to fix in the future.